stage1/iottymux/iottymux.go

```diff
@@ -0,0 +1,573 @@
+			ep := Endpoint{
+				Name:    entry.streamName,
+				Domain:  entry.socketDomain,
+				Address: entry.socketAddress,
+			}
+			streams[i].fifo, err = os.OpenFile(entry.fifoPath, entry.fifoOpenFlags, os.ModeNamedPipe)
+			if err != nil {
+				return fmt.Errorf("invalid %s FIFO: %s", entry.streamName, err)
+			}
+			defer streams[i].fifo.Close()
+			streams[i].listener, err = net.Listen(ep.Domain, ep.Address)
+			if err != nil {
+				return fmt.Errorf("unable to open %s listener: %s", entry.streamName, err)
+			}
+			defer streams[i].listener.Close()
+			endpoints.Targets = append(endpoints.Targets, ep)
+			diag.Printf("Listening for %s on %s\n", entry.streamName, ep.Address)
+		}
+	}
+
+	// write available endpoints to status file
+	sf, err := os.OpenFile(statusFile, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	err = json.NewEncoder(sf).Encode(endpoints)
+	_ = sf.Sync()
+	_ = sf.Close()
+	if err != nil {
+		return err
+	}
+
+	c := make(chan error)
+
+	// TODO(lucab): finalize custom logging modes
+	logMode := os.Getenv("STAGE1_LOGMODE")
+	var logFile *os.File
+	switch logMode {
+	case "k8s-plain":
+		var err error
+		// TODO(lucab): check what should be the target path with Euan
+		logTarget := filepath.Join(pathPrefix, appName, "logfile")
+		logFile, err = os.OpenFile(logTarget, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
+		if err != nil {
+			return err
+		}
+		defer logFile.Close()
+	}
+
+	// proxy stdin
+	if streams[0].fifo != nil && streams[0].listener != nil {
+		clients := make(chan net.Conn)
+		go acceptConn(streams[0].listener, clients, "stdin")
+		go muxInput(clients, streams[0].fifo)
+	}
+
+	// proxy stdout
+	if streams[1].fifo != nil && streams[1].listener != nil {
+		localTargets := make(chan io.WriteCloser)
+		clients := make(chan net.Conn)
+		lines := make(chan []byte)
+		go bufferLine(streams[1].fifo, lines, c)
+		go acceptConn(streams[1].listener, clients, "stdout")
+		go muxOutput("stdout", lines, clients, localTargets)
+		if logFile != nil {
+			localTargets <- logFile
+		}
+	}
+
+	// proxy stderr
+	if streams[2].fifo != nil && streams[2].listener != nil {
+		localTargets := make(chan io.WriteCloser)
+		clients := make(chan net.Conn)
+		lines := make(chan []byte)
+		go bufferLine(streams[2].fifo, lines, c)
+		go acceptConn(streams[2].listener, clients, "stderr")
+		go muxOutput("stderr", lines, clients, localTargets)
+		if logFile != nil {
+			localTargets <- logFile
+		}
+	}
+
+	// If nothing else fails, iomux service will be waiting here forever
+	// and be terminated by systemd only when the main application exits.
+	return <-c
+}
+
+// bufferLine buffers and queues a single line from a Reader to a multiplexer
+// If reading from src fails, it hard-fails and propagates the error back.
+func bufferLine(src io.Reader, c chan<- []byte, ec chan<- error) {
+	rd := bufio.NewReader(src)
+	for {
+		lineOut, err := rd.ReadBytes('\n')
+		if len(lineOut) == 0 && err != nil {
```

Comment:
My understanding is that in this case we got a short read: we flush the incomplete line to the channel and proceed to the next read, until we get an hard failure (len=0 and error).